### PR TITLE
[core] Refactor renderer implementation

### PR DESCRIPTION
**Pull request by pozdnyakov**
This PR is doing two things:
* Merges initialization of render items and render sources within single loop, so that sources will have the layers paint properties all evaluated before the update (this is required for the work on https://github.com/mapbox/mapbox-gl-native/issues/13812 )

* Performance improvements:  around 5% improvements for `API_renderStill_reuse_*` benchmark tests.

### CODE REVIEW

File: src/mbgl/renderer/renderer_impl.cpp [hunk 1]
```diff
@@ -270,6 +236,67 @@ void Renderer::Impl::render(const UpdateParameters& updateParameters) {
...
+            } 
+
+            // Handle layers without source.
+            if (layerNeedsRendering && sourceImpl.get() == sourceImpls->at(0).get()) {            
+                if (parameters.contextMode == GLContextMode::Unique
+                    && layerImpl.get() == layerImpls->at(0).get()) {
+                    const auto& solidBackground = layer->getSolidBackground();
+                    if (solidBackground) {
+                        backgroundColor = *solidBackground;
+                        continue; // This layer is shown with background color, and it shall not be added to render items.
```
Comment (alexshalamov) [inline src/mbgl/renderer/renderer_impl.cpp]: nit: maybe worth adding comment for that, for example `see // - CLEAR block`

File: src/mbgl/tile/geometry_tile.cpp [hunk 2]
```diff
@@ -93,12 +93,13 @@ void GeometryTile::setLayers(const std::vector<Immutable<Layer::Impl>>& layers)
...
     for (const auto& layer : layers) {
         // Skip irrelevant layers.
-        if (layer->getTypeInfo()->source == LayerTypeInfo::Source::NotRequired ||
-            layer->source != sourceID ||
-            id.overscaledZ < std::floor(layer->minZoom) ||
+        assert(layer->getTypeInfo()->source != LayerTypeInfo::Source::NotRequired);
+        assert(layer->source == sourceID);
+        if (id.overscaledZ < std::floor(layer->minZoom) ||
             id.overscaledZ >= std::ceil(layer->maxZoom) ||
             layer->visibility == VisibilityType::None) {
```
Comment (alexshalamov) [inline src/mbgl/tile/geometry_tile.cpp]: nit: visibility is checked in `RenderLayer::needsRendering`, might be converted to assert as well?

File: src/mbgl/renderer/renderer_impl.cpp [hunk 3]
```diff
@@ -289,54 +316,15 @@ void Renderer::Impl::render(const UpdateParameters& updateParameters) {
...
-        if (!source) {
-            Log::Warning(Event::Render, "can't find source for layer '%s'", layer->getID().c_str());
-            continue;
-        }
-
-        layer->setRenderTiles(source->getRenderTiles(), parameters.state);
-        order.emplace_back(*layer, source);
-
-        if (const RenderLayerSymbolInterface* symbolLayer = layer->getSymbolInterface()) {
+        renderItem.layer.setRenderTiles(renderItem.source->getRenderTiles(), parameters.state);
```
Comment (alexshalamov) [inline src/mbgl/renderer/renderer_impl.cpp]: `std::move(renderItem.source->getRenderTiles())`

File: src/mbgl/tile/geometry_tile.cpp (hunk 2 above)
Comment (pozdnyakov) [reply src/mbgl/tile/geometry_tile.cpp]: `needsRendering == true` for a source if there is at least one visible layer, but all the others are also passed -> needs checking :disappointed:  (though the check might be optimized out in future).

File: src/mbgl/renderer/renderer_impl.cpp (hunk 1 above)
Comment (pozdnyakov) [reply src/mbgl/renderer/renderer_impl.cpp]: I'm going to update this code shortly with the follow-up patches for #13812 , so I'd like to defer this nit to that moment.

**Pull request merged**